alerts/alerts.go

```diff
@@ -0,0 +1,244 @@
+
+const (
+	HalfDay  = 60 * 60 * 12
+	Day      = 60 * 60 * 24
+	FiveDays = 60 * 60 * 24 * 5
+)
+
+// Manager manages alerts.
+type Manager interface {
+	// Raise raises an alert.
+	Raise(alert *api.Alert) error
+	// Enumerate lists all alerts filtered by a variadic list of filters.
+	// It will fetch a superset such that every alert is matched by at least one filter.
+	Enumerate(filters ...Filter) ([]*api.Alert, error)
+	// Filter filters given list of alerts successively through each filter.
+	Filter(alerts []*api.Alert, filters ...Filter) ([]*api.Alert, error)
+	// Delete deletes alerts filtered by a chain of filters.
+	Delete(filters ...Filter) error
+	// SetRules sets a set of rules to be performed on alert events.
+	SetRules(rules ...Rule)
+	// DeleteRules deletes rules
+	DeleteRules(rules ...Rule)
+}
+
+func newManager(kv kvdb.Kvdb, options ...Option) (*manager, error) {
+	m := &manager{kv: kv, rules: make(map[string]Rule), ttl: HalfDay}
+	for _, option := range options {
+		switch option.GetType() {
+		case TTLOption:
+			v, ok := option.GetValue().(uint64)
+			if !ok {
+				return nil, typeAssertionError
+			}
+			m.ttl = v
+		}
+	}
+	return m, nil
+}
+
+// manager implements Manager interface.
+type manager struct {
+	kv    kvdb.Kvdb
+	rules map[string]Rule
+	ttl   uint64
+	sync.Mutex
+}
+
+// getKey is a util func that constructs kvdb key.
+// kvdb tree structure is setup as follows:
+// <baseKey>/<resourceType>/<dec2hex(alertType)>/<resourceID>/<alertObject>
+func getKey(resourceType string, alertType int64, resourceID string) string {
+	return filepath.Join(KvdbKey, resourceType, strconv.FormatInt(alertType, 16), resourceID)
+}
+
+func (m *manager) Raise(alert *api.Alert) error {
+	for _, rule := range m.rules {
+		if rule.GetEvent() == RaiseEvent {
+			match, err := rule.GetFilter().Match(alert)
+			if err != nil {
+				return err
+			}
+			if match {
+				if err := rule.GetAction().Run(m); err != nil {
+					return err
+				}
+			}
+		}
+	}
+
+	if alert.Timestamp == nil {
+		alert.Timestamp = &timestamp.Timestamp{Seconds: time.Now().Unix()}
+	}
+
+	if alert.Cleared {
+		// if the alert is marked Cleared, it is pushed to kvdb with a TTLOption of half day
+		_, err := m.kv.Put(getKey(alert.Resource.String(), alert.GetAlertType(), alert.ResourceId), alert, m.ttl)
+		return err
+	} else {
+		_, err := m.kv.Put(getKey(alert.Resource.String(), alert.GetAlertType(), alert.ResourceId), alert, 0)
+		return err
+	}
+}
+
+// Enumerate takes a variadic list of filters that are first analyzed to see if one filter
+// is inclusive of other. Only the filters that are unique supersets are retained and their contents
+// is fetched using kvdb enumerate.
+func (m *manager) Enumerate(filters ...Filter) ([]*api.Alert, error) {
+	myAlerts := make([]*api.Alert, 0, 0)
+	keys, err := getUniqueKeysFromFilters(filters...)
+	if err != nil {
+		return nil, err
+	}
+
+	// enumerate for unique keys
+	for key := range keys {
+		kvps, err := m.kv.Enumerate(key)
```

Comment:
If only ResourceType filter is provided wouldn't this call return only sub trees. In that case you will have again enumerate so that you get all the leafs.

Reply:
done. good point. i added recursive call.